src/dist/replication/common/fs_manager.h

```diff
@@ -91,18 +105,45 @@ class fs_manager
     void update_disk_stat();
 
 private:
+    void reset_disk_stat()
+    {
+        _capacity_total_mb = 0;
+        _available_total_mb = 0;
+        _available_total_ratio = 0;
+        _available_min_ratio = 100;
+        _available_max_ratio = 0;
+    }
+    void compute_disk_density()
+    {
+        for (auto &dir_node : _dir_nodes) {
+            dir_node->set_disk_density(dir_node->disk_available_ratio - _available_total_ratio);
+            _sorted_dir_nodes.emplace(dir_node);
+        }
+    };
     dir_node *get_dir_node(const std::string &subdir);
 
     // when visit the tag/storage of the _dir_nodes map, there's no need to protect by the lock.
     // but when visit the holding_replicas, you must take care.
     mutable zrwlock_nr _lock;
-    std::vector<std::unique_ptr<dir_node>> _dir_nodes;
+
+    int64_t _capacity_total_mb = 0;
+    int64_t _available_total_mb = 0;
+    int64_t _available_total_ratio = 0;
+    int64_t _available_min_ratio = 100;
+    int64_t _available_max_ratio = 0;
+
+    std::vector<std::shared_ptr<dir_node>> _dir_nodes;
+    std::set<std::shared_ptr<dir_node>> _sorted_dir_nodes;
```

Comment:
Elements in `std::set<std::shared_ptr<dir_node>>` are not sorted by `dir_node` but sorted by `std::shared_ptr<>`

Reply:
And `_sorted_dir_nodes ` is only used for test?

Reply:
No, It's for findind the max and min dir_node(sorted by denisty ) for balance

Reply:
> Elements in `std::set<std::shared_ptr<dir_node>>` are not sorted by `dir_node` but sorted by `std::shared_ptr<>`

really? I should test it aganin and modify it. if that, should I only use std::set<dir_node>?

Reply:
This is test only, you'd better remove it.

Reply:
> No, It's for find the max and min dir_node(sorted by denisty ) for balance

Not only for test. And new commit have been modified to use:
```
static bool sorted_by_density(const std::shared_ptr<dir_node> &left,
                                  const std::shared_ptr<dir_node> &right)
    {
        return left->disk_density < right->disk_density;
    };
```
to compare, you can review it.